### PR TITLE
Disconnect before changing cookie

### PR DESCRIPTION
**Pull request by elvanja**
Related to #3 

Notes:
- makes sure that nodes that need to change the cookie are first disconnected from everyone else except the manager
- then the cookie is changed
- and only then the partition is connected back together
- helps with various connection warnings we get otherwise, see related issue

### CODE REVIEW

File: lib/schism.ex [hunk 1]
```diff
@@ -39,20 +39,20 @@ defmodule Schism do
     manager = Node.self()
 
     for node <- nodes do
+      # Force the node to disconnect from all nodes that aren't us
+      all_except_us = :rpc.call(node, Node, :list, []) -- [manager]
+      Enum.each(all_except_us, fn n -> :rpc.call(node, Node, :disconnect, [n]) end)
```
Comment (elvanja) [inline lib/schism.ex:44]: Tried also with `{_, []} = :rpc.multicall(all_except_us, Node, :disconnect, [node])`, without any improvement in final result. Kept this `Enum.each/2` idea but both are OK by me.

File: lib/schism.ex [hunk 2]
```diff
@@ -61,21 +61,20 @@ defmodule Schism do
...
-      for other_node <- nodes do
-        :rpc.call(n, Node, :connect, [other_node])
-      end
-    end
+    # Restore the cookie
+    partition(nodes, Atom.to_string(:erlang.get_cookie()))
   end
 
+  defp connect_nodes([node | other_nodes]) do
+    Enum.each(other_nodes, fn n -> :rpc.call(node, Node, :connect, [n]) end)
```
Comment (elvanja) [inline lib/schism.ex:69]: Similar to disconnect, `{_, []} = :rpc.multicall(other_nodes, Node, :connect, [node])` can be used too.

File: lib/schism.ex [hunk 3]
```diff
@@ -39,20 +39,20 @@ defmodule Schism do
...
-      |> :rpc.call(Node, :list, [])
-      |> Enum.reject(fn n -> n == manager || n in nodes end)
-      |> Enum.each(fn n -> :rpc.call(node, Node, :disconnect, [n]) end)
-
       # Ensure we can still talk to the node
       :pong = Node.ping(node)
     end
 
+    # Reconnect the nodes in partition now that the cookie is the same
+    connect_nodes(nodes)
```
Comment (elvanja) [inline lib/schism.ex:54]: A small optimisation vs:
```elixir
for node <- nodes do
  Enum.each(nodes -- [node], fn n -> :rpc.call(node, Node, :connect, [n]) end)
  # or {_, []} = :rpc.multicall(nodes -- [node], Node, :connect, [node])
end
```

**Review by elvanja — COMMENTED** [review]
All the versions (original, the one in this PR and the one with `:rpc.multicall/4` all exhibit the same problem with following test:

```elixir
test "flutter test" do
  nodes = LocalCluster.start_nodes("flutter-nodes", 5)
  [n1, n2, n3, n4, n5] = nodes

  Enum.each(1..10, fn _ ->
    Schism.partition([n1, n2, n5])
    Schism.partition([n3, n4])
    Schism.heal([n1, n2, n5])
    Schism.heal([n3, n4])
    # or Schism.heal(nodes)
  end)
end
```
Warnings / errors that occur:
```
15:49:26.585 [error] ** Connection attempt from disallowed node :"partition-nodes3@127.0.0.1" **
15:48:25.847 [error] ** Boot server could not send result to socket: {:error, :closed}**
```
However, it occurs far less in this PR than in the original.
Didn't put the above test in the PR since it makes no sense, it would just show that there are those errors/warnings.

**Review by keathley — APPROVED** [review]

**Comment by keathley** [comment]
Thanks! I suspect we'll never be able to get rid of all of the errors or warning messages. We don't control enough of the network stack to make it deterministic. Too many things could be trying to initiate connections in the network. But this should help cut down on the noise and I think it does a better job of describing intent.

**Pull request merged**